### ISSUE TITLE
`@JS()` doesn't support or fail on `external List<NotDynamic> get`

### ISSUE THREAD

**Issue by jakemac53** [dart-lang/sdk#34195]
Original issue file by @jackd https://github.com/dart-lang/build/issues/1771

Dart2js produces code throwing `TypeError`s when using wrapped javascript which is known to return typed arrays. ddc code runs fine. `cast` on the output of dart2js fixes the issue, but is somewhat unsatisfying.

Repo demonstrating the issue [here](https://github.com/jackd/list_cast) (including all code below).

Result of running ddc js:
`woof, moo`

Result of running dart2js js:
`Uncaught Error: TypeError: Closure 'minified:by': type '(minified:F) => String' is not a subtype of type '(dynamic) => String'`

farm.js
```javascript
function Animal(sound) {
  this._sound = sound
}

Animal.prototype.hello = function() {
  return this._sound;
}

function Farm() {
  this._animals = [new Animal('woof'), new Animal('moo')];
}

Farm.prototype.animals = function() {
  return this._animals;
}
```
farm.dart
```dart
@JS()
library farm;

import 'package:js/js.dart';

@JS()
class Animal {
  external factory Animal(String sound);
  external String hello();
}

@JS()
class Farm {
  external factory Farm();
  external List<Animal> animals();
}
```
index.dart
```dart
import 'farm.dart';

void main() {
  var animals = new Farm().animals();
  // animals = animals.cast<Animal>();  // fixes errors in dart2js
  print(animals.map<String>((a) => a.hello()).join(', '));
}
```

index.html
```
<!DOCTYPE html>
<html>
<head>
    <script src="./farm.js"></script>
    <script defer src="./index.dart.js"></script>
</head>
<body></body>
</html>
```

**Comment by jakemac53** [comment]
@jacob314 @rakudrama

**Comment by matanlurey** [comment]
The following code is invalid in Dart 2:

```dart
@JS()
class Farm {
  external factory Farm();
  external List<Animal> animals();
  //                   ^^^^^^^^^
}
```

In JavaScript-compiled Dart, the backing implementation of `List` is `Array`, which of course, does not have generic type arguments - so it is considered a `List<dynamic>` in Dart2JS and DDC. In Dart 1 the following was valid:

```dart
List<Animal> animals = <dynamic>[];
```

This of course, is not valid in Dart 2.

We'll have to either:

* Make it a compile-time error or warning to assume reified generic types on `@JS()` interfaces.
* Add automatic `List.from` or `<List>.cast`'s to all generic lists on `@JS()` interfaces.

You can see some documentation here:
https://github.com/matanlurey/dart_js_interop#generic-type-arguments

**Comment by jackd** [comment]
Thanks for the documentation repo - I can confirm using `--omit-implicit-checks` resolves the issue.

Can I ask what the argument is for the first option over the second? As I understand it, there's no way of getting a typed list back from javascript, so `external List<Animal> animals()` unambiguously communicates a desire to cast the returned value. Are there implementations of casting that are better in certain circumstances than others? If there are, whatever the default behaviour is, it could always be turned off by leaving off the generic types in the wrapper.

I'd rather not have to write
```
class WrappedFarm {
  final Farm _farm;
  WrappedFarm(this.farm);
  List<Animal> animals() => _farm.animals().cast<Animal>();
  // other generic returned types
```
for every class in something like [this](https://github.com/jackd/threejs-dart-facade/blob/master/lib/src/three.dart), which is itself already a wrapper around another source (even if that wrapper is generated automatically).

**Comment by srawlins** [comment]
I think at this point I would prefer option 2, automatic List.from.

Without this change, if we had to wrap ourselves with `List.from()`, I think the best way to write js-interop APIs is like so:

```dart

@JS('foo.bar.getStuff')
external List<dynamic /* num */> _getStuff();
List<num> getStuff() => _getStuff();

@JS('foo.bar.getThings')
external List<dynamic /* num */ > _getThings(Something a, [String b, String c]);

List<num> getThings(Something a, [String b, String c]) {
  if (b == null) {
    return _getThings(a);
  } else {
    return c == null ? _getThings(a, b) : _getThings(a, b, c);
  }
}
```

What a joy that is! (The proper arity is required for `undefined` checks in JS, correct?)

And I'm not sure how you would write a similar wrapper for a class:

```dart
@JS('foo.bar.Bar')
class Bar {
  external factory Bar();
  // Can you wrap this with `@JS()` as well??? 
  // So that you could write this as _getList()?
  external List<num> getList();
}
```

**Comment by srawlins** [comment]
Ah, I see @jackd actually already gave a little solution to the class issue. Here's a slightly different take, with some visibility benefits. Any code outside this library only sees the one class, and is unaware that it's a wrapper. Let's say Farm's constructor also takes a `String name`:

```dart
@JS('Farm')
class _Farm {
  external factory _Farm(String name);
  external List<Animal> animals();
}

class Farm {
  final _Farm _jsFarm;

  factory Farm(String name) {
    var jsFarm = _Farm(name);
    return Farm._(jsFarm);
  }

  Farm._(this._jsFarm);

  List<Animal> animals() => List<Animal>.from(_jsFarm.animals());
}
```

**Comment by matanlurey** [comment]
To be clear, both DDC and Dart2JS always throw. Dart2JS with `--omit-implicit-checks` is not intended to fix bugs, but rather to be used for well-formed applications that work _without_ `--omit-implicit-checks`.

**Comment by jackd** [comment]
Confirmed, sorry for the confusion. That comment was based on my experience with 2.0 dev release.

**Comment by jmesserly** [comment]
I'll add a note to https://github.com/dart-lang/sdk/issues/35084 so we handle this correctly, and merge this bug into that one. Thanks so much for reporting! I'm working on (what I hope will be) big improvements to JS interop.

TL;DR -- I want to support the initial example. Arrays created in JS will not have a Dart reified type, so we'll allow them to be cast to any list type.
